### PR TITLE
[SPARK-10180] [SQL] JDBC datasource are not processing EqualNullSafe filter

### PR DESCRIPTION
**Pull request by HyukjinKwon**
This PR is followed by https://github.com/apache/spark/pull/8391.
Previous PR fixes JDBCRDD to support null-safe equality comparison for JDBC datasource. This PR fixes the problem that it can actually return null as a result of the comparison resulting error as using the value of that comparison.

### CODE REVIEW

**Comment by JoshRosen** [comment]
Please update the pull request description to describe the changes in this patch.

**Comment by JoshRosen** [comment]
Will this case be covered by the end-to-end docker tests?

**Comment by HyukjinKwon** [comment]
Should I better write a test code for this?

**Comment by HyukjinKwon** [comment]
Should I better write some test codes for this?

**Comment by yhuai** [comment]
Can we add a test?

**Comment by HyukjinKwon** [comment]
I added a simple test for this. I wanted to add a test including `null` but when the given value is `null`, Spark converts it to `IsNull`. To make this worse, more complicated logics cannot be applied here to test null-safety properly since now filters are pushed down only for basic operations (not even cast or arithmetic calculations). For now, I think this is the best to test this.

**Comment by HyukjinKwon** [comment]
test this please

**Comment by HyukjinKwon** [comment]
It looks Jenkins does not run the test for the past commits that I made as a user not added to whitelist. Would anybody please run the test for this please if it looks good?

**Comment by yhuai** [comment]
test this please

**Comment by HyukjinKwon** [comment]
Just a question. Now it looks the PR for the end-to-end docker tests is merged. Do you think it needs all the tests for all the databases (namely  `MySQLIntegrationSuite` and `PostgresIntegrationSuite` for now)?

**Comment by rxin** [comment]
@zsxwing can you review this one?

**Comment by maropu** [comment]
@HyukjinKwon +1, I think it'd better to add tests in `MySQLIntegrationSuite` and `PostgresIntegrationSuite`.

**Comment by HyukjinKwon** [comment]
Hm.. actually I think adding tests more in docker is a bit over-tested. It looks the comparison operators I used are all already being used in `compileFilter` and i think the logical operators can be covered in the existing test.

**Comment by maropu** [comment]
Yes and you're right. We need to clearly define the rule that decides which tests should be added in `docker-integration-tests`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala [hunk 1]
```diff
@@ -191,6 +191,9 @@ private[sql] object JDBCRDD extends Logging {
    */
   private def compileFilter(f: Filter): String = f match {
     case EqualTo(attr, value) => s"$attr = ${compileValue(value)}"
+    case EqualNullSafe(attr, value) =>
+      s"(NOT ($attr != ${compileValue(value)} OR $attr IS NULL OR " +
```
Comment (zsxwing) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala]: It took me several minutes to understand it. Maybe `(attr IS NULL AND value IS NULL) OR (attr <> null AND value <> null AND attr = value)` is easier.

**Comment by zsxwing** [comment]
I'm wondering if it's okey to add `sealed` to `Filter` and remove `case _`. If so, the compiler can help us find such issue.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala (hunk 1 above)
Comment (zsxwing) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala]: Ah, sorry, my previous expr doesn't work.

**Comment by HyukjinKwon** [comment]
@zsxwing I will anyway resolve the conflicts first

**Comment by rxin** [comment]
Sorry I kept merging pull requests that made your life harder. Can you bring it up to date again? I promise this is the last one.

**Comment by HyukjinKwon** [comment]
@rxin Nothing is eazy and happy new year. I just resolved conflicts.

**Comment by HyukjinKwon** [comment]
test this please

**Comment by HyukjinKwon** [comment]
Could anybody please type "test this please" for this PR? I can't trigger a test for this PR.

**Comment by HyukjinKwon** [comment]
test this please

**Comment by HyukjinKwon** [comment]
test this please

**Comment by rxin** [comment]
Thanks - I've merged this.

**Pull request closed without merging**